nimare/generate.py

```diff
@@ -0,0 +1,269 @@
+    foci_noise : :obj:`int` or :obj:`list`
+        Number of foci considered to be noise in each study
+        or a list of integers representing how many noise focis
+        there should be in each study.
+    foci_weights : :obj:`list` (Optional)
+        Weighing of each foci representing the probability of
+        that foci being sampled in a study.
+    rng : :class:`numpy.random.RandomState` (Optional)
+        Random state to reproducibly initialize random numbers.
+    sample_size_variance : :obj:`int` (Optional)
+        Variance of the number of participants in each study.
+        (must be >=0).
+
+    Returns
+    -------
+    ground_truth_foci : :obj:`list`
+        generated foci in xyz (mm) coordinates
+    dataset : :class:`nimare.Dataset`
+    """
+    if rng is None:
+        rng = np.random.RandomState(seed=1939)
+    if isinstance(sample_size, list):
+        sample_sizes = sample_size
+        if len(sample_sizes) != studies:
+            raise ValueError("sample_size must be the same length as studies")
+    else:
+        if sample_size_variance is None:
+            raise ValueError("sample_size_variance must be specified")
+        sample_size_lower_limit = int(sample_size - sample_size_variance)
+        # add 1 since randint upper limit is a closed interval
+        sample_size_upper_limit = int(sample_size + sample_size_variance + 1)
+        sample_sizes = rng.randint(sample_size_lower_limit, sample_size_upper_limit, size=studies)
+    ground_truth_foci, foci_dict = create_foci(
+        foci_num, fwhm, studies, foci_coords, foci_noise, foci_weights, rng=rng, space="MNI"
+    )
+
+    source_dict = create_source(foci_dict, sample_sizes)
+    dataset = Dataset(source_dict)
+
+    return ground_truth_foci, dataset
+
+
+def create_source(foci, sample_sizes, space="MNI"):
+    """Create dictionary according to nimads(ish) specification
+
+    Parameters
+    ----------
+    foci : :obj:`dict`
+        A dictionary of foci in xyz (mm) coordinates whose keys represent
+        different studies.
+    sample_sizes : :obj:`list`
+        The sample size for each study
+    space : :obj:`str` (Default="MNI")
+        The template space the coordinates are reported in
+
+    Returns
+    -------
+    source : :obj:`dict`
+        study information in nimads format
+    """
+    source = {}
+    for sample_size, (study, study_foci) in zip(sample_sizes, foci.items()):
+        source[f"study-{study}"] = {
+            "contrasts": {
+                "1": {
+                    "coords": {
+                        "space": space,
+                        "x": [c[0] for c in study_foci],
+                        "y": [c[1] for c in study_foci],
+                        "z": [c[2] for c in study_foci],
+                    },
+                    "metadata": {
+                        "sample_sizes": [sample_size],
+                    },
+                }
+            }
+        }
+
+    return source
+
+
+def create_foci(
```

Comment:
I do a lot of type checking here, I'm curious if there is any preferred way to validate inputs to a function.

Reply:
If there's one valid type for each input we could use `typing`, since NiMARE supports Python 3.6+.

Reply:
sadly I've made it more complicated than that, there are multiple valid types for inputs.

Reply:
Yes, that is also my general strategy throughout the rest of NiMARE.